File: js/src/experimental/anthropic/context.ts
import type { RunTree, RunTreeConfig } from "../../run_trees.js";
import {
  convertFromAnthropicMessage,
  isTaskTool,
  isToolBlock,
  mergeMessagesById,
} from "./messages.js";
import { getCurrentRunTree } from "../../traceable.js";
import {
  aggregateUsageFromModelUsage,
  correctUsageFromResults,
  extractUsageFromMessage,
} from "./usage.js";
import { readTranscript, type TranscriptAssistantTurn } from "./transcripts.js";
import type { SDKMessage, SDKResultMessage } from "./types.js";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value != null && !Array.isArray(value);
}

function isToolResultError(value: unknown): boolean {
  return isRecord(value) && (value.is_error === true || value.isError === true);
}

function makeSubagentTranscriptPathKey(
  path: string,
  toolUseId?: string,
  agentType?: string,
): string {
  return JSON.stringify([path, toolUseId ?? null, agentType ?? null]);
}

/**
 * @internal
 */
export class StreamManager {
  private static liveManagers: Set<StreamManager> = new Set();
  private static managersByRootRun = new WeakMap<RunTree, StreamManager>();

  private rootRun?: RunTree;
  namespaces: { [namespace: string]: RunTree | undefined };
  history: { [namespace: string]: SDKMessage[] };

  assistant: { [messageId: string]: RunTree | undefined } = {};
  tools: { [toolUseId: string]: RunTree | undefined } = {};
  subagents: { [toolUseId: string]: RunTree | undefined } = {};

  mainTranscriptPath?: string;
  subagentTranscriptPaths: {
    path: string;
    toolUseId?: string;
    agentType?: string;
  }[] = [];
  private pendingAgentTools: Map<string, Record<string, unknown>> = new Map();
  private agentToToolUseId: Map<string, string> = new Map();
  private transcriptPathKeys: Set<string> = new Set();
  private resultModelUsage?: SDKResultMessage["modelUsage"];

  postRunQueue: Promise<void>[] = [];
  runTrees: RunTree[] = [];

  constructor() {
    const rootRun = getCurrentRunTree(true);
    this.rootRun = rootRun;
    this.namespaces = rootRun?.createChild ? { root: rootRun } : {};
    this.history = { root: [] };
    if (rootRun != null) {
      StreamManager.managersByRootRun.set(rootRun, this);
    }
    StreamManager.liveManagers.add(this);
  }

  dispose() {
    StreamManager.liveManagers.delete(this);
    if (this.rootRun != null) {
      StreamManager.managersByRootRun.delete(this.rootRun);
    }
  }

  async addMessage(message: SDKMessage) {
    const eventTime = Date.now();

    // Short-circuit if no root run found
    // This can happen if tracing is disabled globally
    if (this.namespaces["root"] == null) return;

    if (message.type === "result") {
      if (message.modelUsage) {
        this.resultModelUsage = message.modelUsage;
      }

      const usage = message.modelUsage
        ? aggregateUsageFromModelUsage(message.modelUsage)
        : extractUsageFromMessage(message);

      if (message.total_cost_usd != null && usage != null) {
        usage.total_cost = message.total_cost_usd;
      }

      this.namespaces["root"].extra ??= {};
      this.namespaces["root"].extra.metadata ??= {};
      this.namespaces["root"].extra.metadata.usage_metadata = usage;

      this.namespaces["root"].extra.metadata.is_error = message.is_error;
      this.namespaces["root"].extra.metadata.num_turns = message.num_turns;
      this.namespaces["root"].extra.metadata.session_id = message.session_id;
      this.namespaces["root"].extra.metadata.duration_ms = message.duration_ms;
      this.namespaces["root"].extra.metadata.duration_api_ms =
        message.duration_api_ms;
    }

    // Skip non-user / non-assistant messages
    if (!("message" in message)) return;

    const namespace = (() => {
      if ("parent_tool_use_id" in message)
        return message.parent_tool_use_id ?? "root";
      return "root";
    })();

    // `eventTime` records the time we receive an event, which for `includePartialMessages: false`
    // equals to the end time of an LLM block, so we need to use the first available end time within namespace.
    const candidateStartTime =
      this.namespaces[namespace]?.child_runs?.at(-1)?.end_time ??
      this.namespaces[namespace]?.start_time ??
      eventTime;

    this.history[namespace] ??= this.history["root"].slice();

    if (message.type === "assistant") {
      const messageId = message.message.id;

      this.assistant[messageId] ??= this.createChild(namespace, {
        name: "claude.assistant.turn",
        run_type: "llm",
        start_time: candidateStartTime,
        inputs: {
          messages: convertFromAnthropicMessage(this.history[namespace]),
        },
        outputs: { output: { messages: [] } },
      });

      if (this.assistant[messageId] == null) return;

      this.assistant[messageId].outputs = (() => {
        const prevMessages =
          this.assistant[messageId].outputs?.output.messages ?? [];
        const newMessages = convertFromAnthropicMessage([message]);
        return {
          output: {
            messages: mergeMessagesById(prevMessages, newMessages),
          },
        };
      })();

      this.assistant[messageId].end_time = eventTime;

      this.assistant[messageId].extra ??= {};
      this.assistant[messageId].extra.metadata ??= {};

      if (message.message.model != null) {
        this.assistant[messageId].extra.metadata.ls_model_name =
          message.message.model;
      }

      this.assistant[messageId].extra.metadata.usage_metadata =
        extractUsageFromMessage(message);

      const tools = Array.isArray(message.message.content)
        ? message.message.content.filter((block) => isToolBlock(block))
        : [];

      for (const block of tools) {
        if (isTaskTool(block)) {
          this.createAgentToolRun(namespace, block, eventTime);
        } else {
          this.createToolRun(namespace, block, eventTime);
        }
      }
    }

    if (message.type === "user") {
      const toolResultBlocks = Array.isArray(message.message.content)
        ? message.message.content.filter((block) => "tool_use_id" in block)
        : [];

      const getToolOutput = (result: unknown) => {
        if (isRecord(result)) {
          return result;
        }

        return { content: result };
      };

      const getToolError = (result: unknown): string => {
        if (["string", "number", "boolean"].includes(typeof result)) {
          return String(result);
        }
        if (Array.isArray(result)) {
          return result.map(getToolError).join("\n");
        }
        if (isRecord(result)) {
          if (typeof result.error === "string") return result.error;
          if (typeof result.text === "string") return result.text;
          if ("content" in result) return getToolError(result.content);
        }
        return JSON.stringify(result);
      };

      for (const block of toolResultBlocks) {
        const tool = this.tools[block.tool_use_id];
        const subagent = this.subagents[block.tool_use_id];
        if (tool != null || subagent != null) {
          // Previous versions of @anthropic-ai/claude-agent-sdk did provide
          // tool result in `message.tool_use_result`, but at least since 0.2.50 it disappeared,
          // so we rely on the last tool result block instead.
          const result =
            message.tool_use_result != null && toolResultBlocks.length === 1
              ? message.tool_use_result
              : block.content;

          const toolOutput = getToolOutput(result);

          const isError = isToolResultError(block) || isToolResultError(result);

          const toolError = isError ? getToolError(result) : undefined;

          await tool?.end(toolOutput, toolError);
          // Match Python's lifecycle: PostToolUse sets outputs on the
          // subagent chain, but the subagent itself is not ended until after
          // transcript reconciliation. Hidden transcript LLM/tool children can
          // arrive after the Agent/Task tool result, so ending the chain here
          // can make reconciled children appear outside their parent bounds.
          if (subagent != null) {
            subagent.outputs ??= toolOutput;
            subagent.error ??= toolError;
          }
        }
      }
    }

    this.history[namespace].push(message);
  }

  addHookEvent(input: unknown, toolUseId?: string) {
    if (typeof input !== "object" || input == null) return;
    const data = input as Record<string, unknown>;

    if (
      this.mainTranscriptPath == null &&
      typeof data.transcript_path === "string"
    ) {
      this.mainTranscriptPath = data.transcript_path;
    }

    if (
      data.hook_event_name === "PreToolUse" &&
      typeof toolUseId === "string" &&
      (data.tool_name === "Agent" || data.tool_name === "Task")
    ) {
      this.pendingAgentTools.set(
        toolUseId,
        typeof data.tool_input === "object" && data.tool_input != null
          ? (data.tool_input as Record<string, unknown>)
          : {},
      );
      return;
    }

    if (data.hook_event_name === "SubagentStart") {
      const agentId =
        typeof data.agent_id === "string" ? data.agent_id : undefined;
      if (agentId == null) return;

      const agentType =
        typeof data.agent_type === "string" ? data.agent_type : undefined;
      let matchedToolUseId: string | undefined;
      for (const [pendingToolUseId, toolInput] of this.pendingAgentTools) {
        const pendingAgentType =
          typeof toolInput.subagent_type === "string"
            ? toolInput.subagent_type
            : typeof toolInput.agent_type === "string"
            ? toolInput.agent_type
            : undefined;
        if (
          agentType == null ||
          pendingAgentType == null ||
          pendingAgentType === agentType
        ) {
          matchedToolUseId = pendingToolUseId;
          break;
        }
      }
      matchedToolUseId ??= this.pendingAgentTools.keys().next().value;
      if (matchedToolUseId != null) {
        this.agentToToolUseId.set(agentId, matchedToolUseId);
        this.pendingAgentTools.delete(matchedToolUseId);
      }
      return;
    }

    if (data.hook_event_name === "SubagentStop") {
      const transcriptPath =
        typeof data.agent_transcript_path === "string"
          ? data.agent_transcript_path
          : undefined;
      if (!transcriptPath) return;

      const agentId =
        typeof data.agent_id === "string" ? data.agent_id : undefined;
      const agentType =
        typeof data.agent_type === "string" ? data.agent_type : undefined;
      const mappedToolUseId =
        agentId != null ? this.agentToToolUseId.get(agentId) : undefined;
      if (agentId != null) this.agentToToolUseId.delete(agentId);

      this.addSubagentTranscriptPath(
        transcriptPath,
        mappedToolUseId,
        agentType,
      );
    }
  }

  addSubagentTranscriptPath(
    path: string,
    toolUseId?: string,
    agentType?: string,
  ) {
    const key = makeSubagentTranscriptPathKey(path, toolUseId, agentType);
    if (this.transcriptPathKeys.has(key)) return;
    this.transcriptPathKeys.add(key);
    this.subagentTranscriptPaths.push({ path, toolUseId, agentType });
  }

  static getActiveToolRun(
    toolName?: string,
    input?: unknown,
  ): RunTree | undefined {
    const currentRun = getCurrentRunTree(true);
    const currentManager =
      currentRun != null
        ? StreamManager.managersByRootRun.get(currentRun)
        : undefined;
    const currentRunTree = currentManager?.getActiveToolRun(toolName, input);
    if (currentRunTree != null) return currentRunTree;

    // Last resort: the SDK invoked an MCP handler from a detached async context
    // that did not inherit the existing LangSmith AsyncLocalStorage. Require
    // both tool name and input to match before scanning live managers to avoid
    // cross-query attribution.
    if (toolName == null || input === undefined) return undefined;

    for (const manager of Array.from(StreamManager.liveManagers).reverse()) {
      if (manager === currentManager) continue;
      const runTree = manager.getActiveToolRun(toolName, input);
      if (runTree != null) return runTree;
    }
    return undefined;
  }

  private getActiveToolRun(
    toolName?: string,
    input?: unknown,
  ): RunTree | undefined {
    const toolEntries = Object.values(this.tools).filter(
      (runTree): runTree is RunTree =>
        runTree != null && runTree.end_time == null && runTree.error == null,
    );

    return toolEntries.find((runTree) => {
      if (toolName != null) {
        const runName = String(runTree.name);
        const nameMatches =
          runName === toolName ||
          runName.includes(toolName) ||
          toolName.includes(runName);
        if (!nameMatches) return false;
      }

      if (input !== undefined) {
        const recorded = runTree.inputs?.input ?? {};
        try {
          return JSON.stringify(recorded) === JSON.stringify(input ?? {});
        } catch {
          return false;
        }
      }

      return true;
    });
  }

  protected createChild(
    namespace: string,
    args: Parameters<RunTree["createChild"]>[0],
  ): RunTree | undefined {
    const parentRunTree = this.namespaces[namespace];
    if (parentRunTree == null) return undefined;
    return this.createChildRun(parentRunTree, args);
  }

  private createChildRun(
    parentRunTree: RunTree,
    args: Parameters<RunTree["createChild"]>[0],
  ): RunTree | undefined {
    const runTree = parentRunTree.createChild(args);
    if (runTree == null) return undefined;

    this.postRunQueue.push(runTree.postRun());
    this.runTrees.push(runTree);
    return runTree;
  }

  private getAgentName(block: { input: Record<string, unknown> }): string {
    const subagentType = block.input.subagent_type;
    if (typeof subagentType === "string" && subagentType.length > 0) {
      return subagentType;
    }

    const agentType = block.input.agent_type;
    if (typeof agentType === "string" && agentType.length > 0) {
      return agentType;
    }

    const description = block.input.description;
    if (typeof description === "string" && description.length > 0) {
      return description.split(" ")[0] || "unknown-agent";
    }

    return "unknown-agent";
  }

  private createToolRun(
    namespace: string,
    block: Record<string, unknown>,
    startTime?: number,
  ): RunTree | undefined {
    if (typeof block.id !== "string") return undefined;

    const name = typeof block.name === "string" ? block.name : "unknown-tool";
    this.tools[block.id] ??=
      this.createChild(namespace, {
        name,
        run_type: "tool",
        inputs: block.input ? { input: block.input } : {},
        start_time: startTime,
      }) ?? this.tools[block.id];
    return this.tools[block.id];
  }

  private createAgentToolRun(
    namespace: string,
    block: Record<string, unknown>,
    startTime?: number,
  ) {
    if (typeof block.id !== "string") return;
    if (
      typeof block.input !== "object" ||
      block.input == null ||
      Array.isArray(block.input)
    ) {
      return;
    }

    const input = block.input as Record<string, unknown>;
    const agentToolRun = this.createToolRun(namespace, block, startTime);
    if (agentToolRun == null) return;

    this.subagents[block.id] ??=
      this.createChildRun(agentToolRun, {
        name: this.getAgentName({ input }),
        run_type: "chain",
        inputs: input,
        start_time: startTime,
        extra: {
          metadata: {
            ls_agent_type: "subagent",
          },
        },
      }) ?? this.subagents[block.id];

    this.namespaces[block.id] ??= this.subagents[block.id];
  }

  private resolveSubagentNamespace(agentType?: string): string | undefined {
    const entries = Object.entries(this.namespaces);
    if (agentType == null) return undefined;
    return entries.find(([, runTree]) => runTree?.name === agentType)?.[0];
  }

  private createSyntheticAssistantRun(
    namespace: string,
    turn: TranscriptAssistantTurn,
  ) {
    let runTree = this.assistant[turn.messageId];

    if (runTree == null) {
      runTree = this.createChild(namespace, {
        name: "claude.assistant.turn",
        run_type: "llm",
        start_time: turn.timestamp,
        inputs:
          turn.inputMessages.length > 0 ? { messages: turn.inputMessages } : {},
        outputs: {
          output: { messages: convertFromAnthropicMessage(turn.message) },
        },
        extra: {
          metadata: {
            ls_provider: "anthropic",
            ...(turn.model != null ? { ls_model_name: turn.model } : {}),
            ...(turn.usageMetadata != null
              ? { usage_metadata: turn.usageMetadata }
              : {}),
          },
        },
      });
      if (runTree == null) return;
      this.assistant[turn.messageId] = runTree;
    } else {
      runTree.outputs = {
        output: { messages: convertFromAnthropicMessage(turn.message) },
      };
      runTree.extra ??= {};
      runTree.extra.metadata ??= {};
      if (turn.model != null) runTree.extra.metadata.ls_model_name = turn.model;
      if (turn.usageMetadata != null) {
        runTree.extra.metadata.usage_metadata = turn.usageMetadata;
      }
    }

    runTree.end_time = turn.timestamp;

    const tools = Array.isArray(turn.message.message.content)
      ? turn.message.message.content.filter((block) => isToolBlock(block))
      : [];

    for (const block of tools) {
      if (isTaskTool(block)) {
        this.createAgentToolRun(namespace, block, turn.timestamp);
      } else {
        this.createToolRun(namespace, block, turn.timestamp);
      }
    }
  }

  private async reconcileTranscripts() {
    const usageByMessageId: Record<string, Record<string, unknown>> = {};

    if (this.mainTranscriptPath != null) {
      const transcript = await readTranscript(this.mainTranscriptPath);
      Object.assign(usageByMessageId, transcript.usageByMessageId);
    }

    for (const transcriptPath of this.subagentTranscriptPaths) {
      const namespace =
        transcriptPath.toolUseId ??
        this.resolveSubagentNamespace(transcriptPath.agentType);
      if (namespace == null || this.namespaces[namespace] == null) continue;

      const transcript = await readTranscript(transcriptPath.path);
      Object.assign(usageByMessageId, transcript.usageByMessageId);

      for (const turn of transcript.turns) {
        this.createSyntheticAssistantRun(namespace, turn);
      }

      for (const toolResult of transcript.toolResults) {
        const tool = this.tools[toolResult.toolUseId];
        if (tool == null) continue;
        const output = isRecord(toolResult.content)
          ? toolResult.content
          : { content: toolResult.content };
        const error = toolResult.isError
          ? typeof toolResult.content === "string" ||
            typeof toolResult.content === "number" ||
            typeof toolResult.content === "boolean"
            ? String(toolResult.content)
            : JSON.stringify(toolResult.content)
          : undefined;
        await tool.end(output, error);
      }
    }

    for (const [messageId, usage] of Object.entries(usageByMessageId)) {
      const runTree = this.assistant[messageId];
      if (runTree == null) continue;
      runTree.extra ??= {};
      runTree.extra.metadata ??= {};
      runTree.extra.metadata.usage_metadata = usage;
    }
  }

  async finish() {
    try {
      await this.reconcileTranscripts();

      if (this.resultModelUsage != null) {
        correctUsageFromResults(
          this.resultModelUsage,
          Object.values(this.assistant).filter((runTree) => runTree != null),
        );
      }

      // Clean up incomplete tools and finalise subagent calls. This mirrors the
      // Python integration: Agent/Task tool runs are ended when their tool result
      // arrives, while subagent chain runs are finalised only after transcript
      // reconciliation so hidden child LLM/tool runs are created first.
      for (const tool of Object.values(this.tools)) {
        if (tool == null) continue;
        if (tool.outputs == null && tool.error == null) {
          await tool.end(undefined, "Run not completed (conversation ended)");
        }
      }

      for (const subagent of Object.values(this.subagents)) {
        if (subagent == null) continue;
        if (subagent.end_time == null) {
          if (subagent.outputs == null && subagent.error == null) {
            await subagent.end(
              undefined,
              "Run not completed (conversation ended)",
            );
          } else {
            await subagent.end();
          }
        }
      }

      // First make sure all the runs are created
      await Promise.allSettled(this.postRunQueue);

      // Then patch the runs
      await Promise.allSettled(
        this.runTrees.map((runTree) => runTree.patchRun()),
      );
    } finally {
      this.dispose();
    }
  }
}

/**
 * Configuration options for wrapping Claude Agent SDK with LangSmith tracing.
 */
export type WrapClaudeAgentSDKConfig = Partial<
  Omit<
    RunTreeConfig,
    "inputs" | "outputs" | "run_type" | "child_runs" | "parent_run" | "error"
  >
>;

File: python/langsmith/integrations/claude_agent_sdk/_context.py
"""Stream context management for Claude Agent SDK tracing."""

import logging
import time
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any, Optional

from ._messages import (
    convert_from_anthropic_message,
    extract_usage_from_assistant_message,
    is_task_tool,
    is_tool_block,
)
from ._usage import aggregate_usage_from_model_usage, correct_usage_from_results

if TYPE_CHECKING:
    from langsmith.run_trees import RunTree

logger = logging.getLogger(__name__)


class StreamManager:
    """Manages message accumulation and run tree creation for Claude Agent SDK.

    This class replaces TurnLifecycle with a JS-aligned approach that:
    - Accumulates outputs to the same run using message keys
    - Tracks tools by tool_use_id
    - Creates proper parent-child relationships for subagents
    """

    def __init__(self, root_run: "RunTree"):
        """Initialize StreamManager with the root run tree.

        Args:
            root_run: The root RunTree for the conversation
        """
        self.namespaces: dict[str, RunTree] = {"root": root_run}
        self.history: dict[str, list[Any]] = {"root": []}

        # Assistant runs keyed by {namespace}:{model}:{sequence}
        self.assistant_runs: dict[str, RunTree] = {}

        # Tool runs keyed by tool_use_id
        self.tools: dict[str, RunTree] = {}

        # Sequence counters for generating run keys (workaround for missing message ID)
        self._sequence_counters: dict[str, int] = {}

        # Track all run trees for cleanup
        self._run_trees: list[RunTree] = []

        # Track tool_use_ids managed by this StreamManager
        self._managed_tool_ids: set[str] = set()

    def _get_run_key(self, msg: Any, namespace: str) -> str:
        """Generate a unique key for an assistant run.

        Since Python SDK AssistantMessage doesn't have an 'id' field like JS,
        we key by namespace:model:sequence instead.
        """
        model = getattr(msg.message, "model", None) if hasattr(msg, "message") else None
        model = model or "unknown"
        key_base = f"{namespace}:{model}"
        self._sequence_counters.setdefault(key_base, 0)
        self._sequence_counters[key_base] += 1
        return f"{key_base}:{self._sequence_counters[key_base]}"

    def _create_child(
        self,
        namespace: str,
        name: str,
        run_type: str,
        inputs: dict[str, Any],
        start_time: float,
        outputs: Optional[dict[str, Any]] = None,
        extra: Optional[dict[str, Any]] = None,
    ) -> "RunTree":
        """Create a child run under the specified namespace."""
        parent = self.namespaces.get(namespace)
        if not parent:
            parent = self.namespaces["root"]

        run_tree = parent.create_child(
            name=name,
            run_type=run_type,
            inputs=inputs,
            outputs=outputs,
            extra=extra,
            start_time=datetime.fromtimestamp(start_time, tz=timezone.utc),
        )

        try:
            run_tree.post()
        except Exception as e:
            logger.warning(f"Failed to post run {name}: {e}")

        self._run_trees.append(run_tree)
        return run_tree

    def add_message(self, msg: Any, event_time: Optional[float] = None) -> None:
        """Process and add a message to the stream.

        Args:
            msg: SDK message (AssistantMessage, UserMessage, ResultMessage)
            event_time: Time the message was received
        """
        if event_time is None:
            event_time = time.time()

        # Short-circuit if no root run
        if "root" not in self.namespaces:
            return

        msg_type = type(msg).__name__

        if msg_type == "ResultMessage":
            self._handle_result_message(msg)
            return

        # Skip non-user/non-assistant messages
        if msg_type not in ("AssistantMessage", "UserMessage"):
            return

        # Determine namespace from parent_tool_use_id
        parent_tool_use_id = getattr(msg, "parent_tool_use_id", None)
        namespace = parent_tool_use_id if parent_tool_use_id else "root"

        # Calculate candidate start time
        parent_run = self.namespaces.get(namespace, self.namespaces["root"])
        child_runs = parent_run.child_runs or []
        if child_runs:
            candidate_start = (
                child_runs[-1].end_time.timestamp()
                if child_runs[-1].end_time
                else event_time
            )
        elif parent_run.start_time:
            candidate_start = parent_run.start_time.timestamp()
        else:
            candidate_start = event_time

        # Initialize history for this namespace
        if namespace not in self.history:
            self.history[namespace] = self.history["root"].copy()

        if msg_type == "AssistantMessage":
            self._handle_assistant_message(msg, namespace, candidate_start, event_time)
        elif msg_type == "UserMessage":
            self._handle_user_message(msg, namespace, event_time)

        # Add to history
        self.history[namespace].append(msg)

    def _handle_assistant_message(
        self,
        msg: Any,
        namespace: str,
        start_time: float,
        event_time: float,
    ) -> None:
        """Handle AssistantMessage: create or update LLM run, create tool runs."""
        run_key = self._get_run_key(msg, namespace)

        # Create or get assistant run
        if run_key not in self.assistant_runs:
            history_messages = convert_from_anthropic_message(self.history[namespace])
            self.assistant_runs[run_key] = self._create_child(
                namespace=namespace,
                name="claude.assistant.turn",
                run_type="llm",
                inputs={"messages": history_messages},
                start_time=start_time,
                outputs={"output": {"messages": []}},
            )

        run = self.assistant_runs[run_key]

        # Accumulate outputs
        prev_messages = (
            run.outputs.get("output", {}).get("messages", []) if run.outputs else []
        )
        new_messages = convert_from_anthropic_message([msg])
        run.outputs = {"output": {"messages": prev_messages + new_messages}}

        # Update end time
        run.end_time = datetime.fromtimestamp(event_time, tz=timezone.utc)

        # Set model name
        if run.extra is None:
            run.extra = {}
        run.extra.setdefault("metadata", {})

        model = getattr(msg.message, "model", None) if hasattr(msg, "message") else None
        if model:
            run.extra["metadata"]["ls_model_name"] = model

        # Extract usage
        usage = extract_usage_from_assistant_message(msg)
        run.extra["metadata"]["usage_metadata"] = usage

        # Process tool uses
        self._handle_tool_uses(msg, namespace, event_time)

    def _handle_tool_uses(self, msg: Any, namespace: str, event_time: float) -> None:
        """Process tool uses in an AssistantMessage."""
        if not hasattr(msg, "content"):
            return

        parent_tool_use_id = getattr(msg, "parent_tool_use_id", None)

        for block in msg.content:
            if not is_tool_block(block):
                continue

            tool_use_id = getattr(block, "id", None)
            if not tool_use_id:
                continue

            tool_name = getattr(block, "name", "unknown-tool")
            tool_input = getattr(block, "input", {})

            # Mark this tool as managed by StreamManager
            self._managed_tool_ids.add(tool_use_id)

            if is_task_tool(block) and not parent_tool_use_id:
                # This is a Task tool (subagent) at root level
                subagent_name = (
                    tool_input.get("subagent_type")
                    or tool_input.get("agent_type")
                    or (
                        tool_input.get("description", "").split()[0]
                        if tool_input.get("description")
                        else None
                    )
                    or "unknown-agent"
                )

                if tool_use_id not in self.tools:
                    self.tools[tool_use_id] = self._create_child(
                        namespace="root",
                        name=subagent_name,
                        run_type="chain",
                        inputs=tool_input,
                        start_time=event_time,
                    )
                    # Register this tool's run as a namespace for child messages
                    self.namespaces[tool_use_id] = self.tools[tool_use_id]
            else:
                # Regular tool
                if tool_use_id not in self.tools:
                    self.tools[tool_use_id] = self._create_child(
                        namespace=namespace,
                        name=tool_name,
                        run_type="tool",
                        inputs={"input": tool_input} if tool_input else {},
                        start_time=event_time,
                    )

    def _handle_user_message(
        self,
        msg: Any,
        namespace: str,
        event_time: float,
    ) -> None:
        """Handle UserMessage: complete tool runs with results."""
        if not hasattr(msg, "content"):
            return

        # Find tool result blocks
        content = msg.content
        if not isinstance(content, list):
            return

        tool_result = None
        for block in content:
            block_type = type(block).__name__
            if block_type == "ToolResultBlock" or (
                isinstance(block, dict) and block.get("type") == "tool_result"
            ):
                tool_result = block
                break

        if not tool_result:
            return

        # Get tool_use_id
        tool_use_id = (
            tool_result.get("tool_use_id")
            if isinstance(tool_result, dict)
            else getattr(tool_result, "tool_use_id", None)
        )

        if not tool_use_id or tool_use_id not in self.tools:
            return

        # Get tool response
        tool_response = getattr(msg, "tool_use_result", None)
        if tool_response is None:
            # Try to get from block content
            tool_response = (
                tool_result.get("content")
                if isinstance(tool_result, dict)
                else getattr(tool_result, "content", None)
            )

        # Format output
        if isinstance(tool_response, dict):
            outputs = tool_response
        elif isinstance(tool_response, list):
            outputs = {"content": tool_response}
        else:
            outputs = {"output": str(tool_response)} if tool_response else {}

        # Check for error
        is_error = (
            tool_result.get("is_error", False)
            if isinstance(tool_result, dict)
            else getattr(tool_result, "is_error", False)
        )

        error_msg = None
        if is_error:
            if isinstance(tool_response, (str, int, float, bool)):
                error_msg = str(tool_response)
            elif tool_response is not None:
                try:
                    import json

                    error_msg = json.dumps(tool_response)
                except (TypeError, ValueError):
                    error_msg = str(tool_response)

        # Complete the tool run
        tool_run = self.tools[tool_use_id]
        tool_run.end(outputs=outputs, error=error_msg)
        tool_run.end_time = datetime.fromtimestamp(event_time, tz=timezone.utc)

    def _handle_result_message(self, msg: Any) -> None:
        """Handle ResultMessage: apply usage correction and metadata."""
        from ._messages import extract_usage_from_result_message

        # Get usage from modelUsage (per-model breakdown) or usage (aggregate)
        model_usage = getattr(msg, "model_usage", None)
        if model_usage is None:
            model_usage = getattr(msg, "modelUsage", None)

        # Calculate aggregate usage
        if model_usage:
            usage = aggregate_usage_from_model_usage(model_usage)
            # Apply usage correction to distribute tokens across runs
            correct_usage_from_results(
                model_usage,
                list(self.assistant_runs.values()),
            )
        else:
            # Fall back to extracting from msg.usage
            usage = extract_usage_from_result_message(msg)
            # Apply aggregate usage to the last assistant run (like old code)
            if self.assistant_runs and usage:
                last_run = list(self.assistant_runs.values())[-1]
                if last_run.extra is None:
                    last_run.extra = {}
                last_run.extra.setdefault("metadata", {})
                last_run.extra["metadata"]["usage_metadata"] = usage.copy()

        # Calculate aggregate usage for root run
        root_run = self.namespaces.get("root")
        if not root_run:
            return

        if root_run.extra is None:
            root_run.extra = {}
        root_run.extra.setdefault("metadata", {})

        # Add total_cost if available
        total_cost = getattr(msg, "total_cost_usd", None)
        if total_cost is not None:
            usage["total_cost"] = total_cost

        root_run.extra["metadata"]["usage_metadata"] = usage

        # Add conversation-level metadata
        metadata_attrs = (
            "is_error",
            "num_turns",
            "session_id",
            "duration_ms",
            "duration_api_ms",
        )
        for attr in metadata_attrs:
            val = getattr(msg, attr, None)
            if val is not None:
                root_run.extra["metadata"][attr] = val

    def get_final_output(self) -> Optional[dict[str, Any]]:
        """Get the final output from the last assistant message."""
        # Get all assistant run outputs
        for run in reversed(list(self.assistant_runs.values())):
            if run.outputs:
                messages = run.outputs.get("output", {}).get("messages", [])
                if messages:
                    return messages[-1]
        return None

    def get_managed_tool_ids(self) -> set[str]:
        """Get the set of tool_use_ids managed by this StreamManager."""
        return self._managed_tool_ids.copy()

    def finish(self) -> None:
        """Clean up incomplete tools and patch all runs."""
        # End incomplete tool runs
        for tool_use_id, tool_run in self.tools.items():
            if tool_run.outputs is None and tool_run.error is None:
                tool_run.end(error="Run not completed (conversation ended)")

        # Patch all runs
        for run_tree in self._run_trees:
            try:
                run_tree.patch()
            except Exception as e:
                logger.warning(f"Failed to patch run: {e}")
